src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java

```diff
@@ -64,202 +120,3415 @@ public class QuizFacadeIT extends IsaacIntegrationTest {
+          (List<QuizAttemptDTO>) getFreeAttemptsResponse.getEntity();
+      assertTrue(responseBody.isEmpty());
+    }
+
+    @Test
+    public void studentFreeAttempts_oneFreeAttempt()
+        throws NoCredentialsAvailableException, NoUserException, SegueDatabaseException,
+        AuthenticationProviderMappingException, IncorrectCredentialsProvidedException,
+        AdditionalAuthenticationRequiredException, InvalidKeySpecException, NoSuchAlgorithmException,
+        MFARequiredButNotConfiguredException {
+      LoginResult studentLogin = loginAs(httpSession, TEST_STUDENT_BOB_EMAIL, TEST_STUDENT_BOB_PASSWORD);
+      HttpServletRequest getFreeAttemptsRequest = createRequestWithCookies(new Cookie[] {studentLogin.cookie});
+      replay(getFreeAttemptsRequest);
+
+      Response getFreeAttemptsResponse = quizFacade.getFreeAttempts(getFreeAttemptsRequest);
+
+      assertEquals(Response.Status.OK.getStatusCode(), getFreeAttemptsResponse.getStatus());
+
+      @SuppressWarnings("unchecked") List<QuizAttemptDTO> responseBody =
+          (List<QuizAttemptDTO>) getFreeAttemptsResponse.getEntity();
+      assertTrue(responseBody.stream().anyMatch(q -> q.getQuizId().equals(QUIZ_TEST_QUIZ_ID)));
+    }
+  }
+
+  @Nested
+  class GetQuizAssignment {
+    @Nested
+    class MissingOrInvalidData {
+      @Test
+      public void noQuizId() {
+        HttpServletRequest getQuizAssignmentRequest = createNiceMock(HttpServletRequest.class);
+        replay(getQuizAssignmentRequest);
+
+        Response getQuizAssignmentResponse = quizFacade.getQuizAssignment(getQuizAssignmentRequest, null);
+
+        assertEquals(Response.Status.BAD_REQUEST.getStatusCode(), getQuizAssignmentResponse.getStatus());
+
+        assertEquals("You must provide a valid test assignment id.",
+            getQuizAssignmentResponse.readEntity(SegueErrorResponse.class).getErrorMessage());
+      }
+
+      @Test
+      public void cancelledAssignment() throws NoCredentialsAvailableException, NoUserException, SegueDatabaseException,
+          AuthenticationProviderMappingException, IncorrectCredentialsProvidedException,
+          AdditionalAuthenticationRequiredException, InvalidKeySpecException, NoSuchAlgorithmException,
+          MFARequiredButNotConfiguredException {
+        LoginResult teacherLogin = loginAs(httpSession, TEST_TEACHER_EMAIL, TEST_TEACHER_PASSWORD);
+        HttpServletRequest getQuizAssignmentRequest = createRequestWithCookies(new Cookie[] {teacherLogin.cookie});
+        replay(getQuizAssignmentRequest);
+
+        Response getQuizAssignmentResponse =
+            quizFacade.getQuizAssignment(getQuizAssignmentRequest, QUIZ_ASSIGNMENT_CANCELLED_ID);
+
+        assertEquals(Response.Status.BAD_REQUEST.getStatusCode(), getQuizAssignmentResponse.getStatus());
+
+        assertEquals("This assignment has been cancelled.",
+            getQuizAssignmentResponse.readEntity(SegueErrorResponse.class).getErrorMessage());
```

Comment:
Having all these chained calls makes your code hard to read. In cases like this, it's more readable to simply save the result in a variable. For instance, I would have written this assertion like this:
```java
String expectedErrorMessage = "This assignment has been cancelled.";
String actualErrorMessage = getQuizAssignmentResponse.readEntity(SegueErrorResponse.class).getErrorMessage();
assertEquals(expectedErrorMessage, actualErrorMessage);
```

Or even like this:
```java
SegueErrorResponse errorResponse = getQuizAssignmentResponse.readEntity(SegueErrorResponse.class);
String actualErrorMessage = errorResponse.getErrorMessage();
String expectedErrorMessage = "This assignment has been cancelled.";
assertEquals(expectedErrorMessage, actualErrorMessage);
```

That does not mean that you have to do this for every single assertion. When you have short assertions like the following, it's fine, because they're easily readable:
```java
assertEquals(QUIZ_TEST_QUIZ_ID, responseBody.getQuizId());
```